File: madameautomata-bot-python-autoreg/bot/util/perm.py

from .db import Session, Config
from sqlalchemy import select
from discord import Member

async def _get_cfg():
    async with Session() as s:
        row = (await s.execute(select(Config).where(Config.id==1))).scalar_one_or_none()
        return row

async def is_domme(m: Member | None) -> bool:
    if not m: return False
    cfg = await _get_cfg()
    if not cfg or not cfg.domme_role_id: return False
    return any(r.id == cfg.domme_role_id for r in m.roles)

async def is_sub(m: Member | None) -> bool:
    if not m: return False
    cfg = await _get_cfg()
    if not cfg or not cfg.sub_role_id: return False
    return any(r.id == cfg.sub_role_id for r in m.roles)

async def is_staff(m: Member | None) -> bool:
    if not m: return False
    cfg = await _get_cfg()
    if not cfg or not cfg.staff_role_id: return False
    return any(r.id == cfg.staff_role_id for r in m.roles)

File: madameautomata-js-bot/src/util/perm.js
import { prisma } from "../lib/prisma.js";

function hasRoleName(member, names = []) {
  if (!member?.roles?.cache) return false;
  const set = new Set(names.map(n => n.toLowerCase()));
  for (const r of member.roles.cache.values()) {
    if (set.has(r.name.toLowerCase())) return true;
  }
  return false;
}

export async function getConfig() {
  let row = await prisma.config.findUnique({ where: { id: 1 } });
  if (!row) row = await prisma.config.create({ data: { id: 1 } });
  return row;
}

export async function isDomme(member) {
  if (!member) return false;
  const cfg = await getConfig();
  if (cfg.dommeRoleId && member.roles?.cache?.has(cfg.dommeRoleId)) return true;
  if (hasRoleName(member, ["Domme", "Mistress", "Dominant"])) return true;
  if (member.permissions?.has("ManageGuild") || member.permissions?.has("ManageRoles")) return true;
  return false;
}

export async function isSub(member) {
  if (!member) return false;
  const cfg = await getConfig();
  if (cfg.subRoleId && member.roles?.cache?.has(cfg.subRoleId)) return true;
  if (hasRoleName(member, ["Sub", "Submissive", "Pet", "Slave"])) return true;
  return false;
}

export async function isStaff(member) {
  if (!member) return false;
  const cfg = await getConfig();
  if (cfg.staffRoleId && member.roles?.cache?.has(cfg.staffRoleId)) return true;
  if (hasRoleName(member, ["Staff", "Owner", "Admin", "Mod", "Moderator"])) return true;
  if (member.permissions?.has("Administrator")) return true;
  return false;
}
